File: table.py
import re
import textwrap

ANSI_ESCAPE = re.compile(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~]|\].*?\x07|\].*?\x1b\\)')


def strip_ansi(text: str) -> str:
    return ANSI_ESCAPE.sub("", text)



class Table:
    def __init__(self, title: str, columns_data: list = None, width: int = 80):
        self.title = title
        self.columns_data = columns_data or []
        self._total_width = self._calculate_total_width()
        self.width = width


    def _calculate_total_width(self) -> int:
        if not self.columns_data:
            return 0

        total_columns_width = sum(width for _, width in self.columns_data)

        num_spaces = len(self.columns_data) - 1

        return total_columns_width + num_spaces

    def print_header(self):
        print("\n" + self.title.center(self._total_width) + "\n")

        header_parts = []
        separator_parts = []
        for name, width in self.columns_data:
            header_parts.append(f"{name:<{width}}")
            separator_parts.append(f"{'-' * width:<{width}}")

        print(" ".join(header_parts))
        print(" ".join(separator_parts))

    def print_row(self, row_values: list):
        if len(row_values) != len(self.columns_data):
            print(
                f"Error: Row values count ({len(row_values)}) does not match column count ({len(self.columns_data)}).")
            return

        content_parts = []
        for i, (col_name, width) in enumerate(self.columns_data):
            value = str(row_values[i])  # Ensure value is a string
            content_parts.append(f"{value:<{width}}")

        print(" ".join(content_parts))
    def print_row2(self, row_values: list):
        if len(row_values) != len(self.columns_data):
            print(
                f"Error: Row values count ({len(row_values)}) does not match column count ({len(self.columns_data)})."
            )
            return

        wrapped_cells = []
        for i, (col_name, width) in enumerate(self.columns_data):
            value = str(row_values[i])
            wrapped = textwrap.wrap(value, width=width) or [""]
            wrapped_cells.append(wrapped)

        max_lines = max(len(cell) for cell in wrapped_cells)

        for line_idx in range(max_lines):
            line_parts = []
            for col_idx, (col_name, width) in enumerate(self.columns_data):
                cell_lines = wrapped_cells[col_idx]
                if line_idx < len(cell_lines):
                    part = f"{cell_lines[line_idx]:<{width}}"
                else:
                    part = " " * width
                line_parts.append(part)
            print(" ".join(line_parts))
    def print_closing(self):
        separator_parts = []
        for _, width in self.columns_data:
            separator_parts.append(f"{'-' * width:<{width}}")
        print(" ".join(separator_parts))


    def print_vertical(self, data: dict, wrap_keys: int = 20, wrap_desc: bool = True):
        print("\n" + f"{self.title}".center(self.width))
        print("-" * self.width)

        for key, value in data.items():
            if isinstance(value, dict):
                value = value.get("username", "N/A")

            if isinstance(value, list):
                value = ", ".join(map(str, value))

            value = str(value)

            wrapped = textwrap.wrap(value, width=self.width - wrap_keys - 2) or [""]
            print(f"{key:<{wrap_keys}} {wrapped[0]}")
            for cont in wrapped[1:]:
                print(" " * wrap_keys + " " + cont)

        print("-" * self.width)

File: src/main/kotlin/util/Table.kt
package com.magnariuk.util

import kotlin.collections.get
import kotlin.collections.iterator

class Table(
    private val title: String,
    private val columns: List<Pair<String, Int>> = emptyList(),
    private val width: Int = 80
) {
    private val totalWidth: Int = calculateTotalWidth()

    private fun calculateTotalWidth(): Int {
        if (columns.isEmpty()) return 0
        val totalColumnsWidth = columns.sumOf { it.second }
        val numSpaces = columns.size - 1
        return totalColumnsWidth + numSpaces
    }

    fun printHeader() {
        println("\n" + title.center(totalWidth) + "\n")

        val headerParts = mutableListOf<String>()
        val separatorParts = mutableListOf<String>()

        for ((name, colWidth) in columns) {
            headerParts.add(name.padEnd(colWidth))
            separatorParts.add("-".repeat(colWidth).padEnd(colWidth))
        }

        println(headerParts.joinToString(" "))
        println(separatorParts.joinToString(" "))
    }

    fun printRow(rowValues: List<Any?>) {
        if (rowValues.size != columns.size) {
            println("Error: Row values count (${rowValues.size}) does not match column count (${columns.size}).")
            return
        }

        val contentParts = rowValues.mapIndexed { i, value ->
            value.toString().padEnd(columns[i].second)
        }
        println(contentParts.joinToString(" "))
    }

    fun printRowWrapped(rowValues: List<Any?>) {
        if (rowValues.size != columns.size) {
            println("Error: Row values count (${rowValues.size}) does not match column count (${columns.size}).")
            return
        }

        val wrappedCells = rowValues.mapIndexed { i, value ->
            wrapText(value.toString(), columns[i].second)
        }

        val maxLines = wrappedCells.maxOf { it.size }

        for (lineIdx in 0 until maxLines) {
            val lineParts = wrappedCells.mapIndexed { colIdx, cellLines ->
                if (lineIdx < cellLines.size) {
                    cellLines[lineIdx].padEnd(columns[colIdx].second)
                } else {
                    " ".repeat(columns[colIdx].second)
                }
            }
            println(lineParts.joinToString(" "))
        }
    }

    fun printClosing() {
        val separatorParts = columns.map { "-".repeat(it.second).padEnd(it.second) }
        println(separatorParts.joinToString(" "))
    }

    fun printVertical(data: Map<String, Any?>, wrapKeys: Int = 20) {
        println("\n" + title.center(width))
        println("-".repeat(width))

        for ((key, rawValue) in data) {
            val value = when (rawValue) {
                is Map<*, *> -> rawValue["username"]?.toString() ?: "N/A"
                is List<*> -> rawValue.joinToString(", ")
                else -> rawValue?.toString() ?: "N/A"
            }

            val wrapped = wrapText(value, width - wrapKeys - 2)
            println(key.padEnd(wrapKeys) + " " + wrapped.first())
            for (line in wrapped.drop(1)) {
                println(" ".repeat(wrapKeys) + " " + line)
            }
        }

        println("-".repeat(width))
    }

    private fun wrapText(text: String, lineWidth: Int): List<String> {
        if (text.isEmpty()) return listOf("")
        val words = text.split(" ")
        val lines = mutableListOf<String>()
        var currentLine = StringBuilder()

        for (word in words) {
            if (currentLine.isEmpty()) {
                currentLine.append(word)
            } else if (currentLine.length + 1 + word.length <= lineWidth) {
                currentLine.append(" ").append(word)
            } else {
                lines.add(currentLine.toString())
                currentLine = StringBuilder(word)
            }
        }
        if (currentLine.isNotEmpty()) lines.add(currentLine.toString())
        return lines
    }

    private fun String.center(width: Int): String {
        if (this.length >= width) return this
        val padding = width - this.length
        val left = padding / 2
        val right = padding - left
        return " ".repeat(left) + this + " ".repeat(right)
    }
}